src/coreclr/classlibnative/bcltype/arraynative.cpp

```diff
@@ -1012,81 +1008,71 @@ FCIMPL4(void, ArrayNative::GetReference, ArrayBase* refThisUNSAFE, TypedByRef* e
         if (elemtype == ELEMENT_TYPE_PTR || elemtype == ELEMENT_TYPE_FNPTR)
             FCThrowResVoid(kNotSupportedException, W("NotSupported_Type"));
     }
-#ifdef _DEBUG
-    CorElementType elemtype = arrayElementType.GetInternalCorElementType();
-    _ASSERTE(elemtype != ELEMENT_TYPE_PTR && elemtype != ELEMENT_TYPE_FNPTR);
-#endif
 
-    elemRef->data = refThis->GetDataPtr() + (Offset * refThis->GetComponentSize());
-    elemRef->type = arrayElementType;
-}
-FCIMPLEND
-
-FCIMPL2(void, ArrayNative::SetValue, TypedByRef * target, Object* objUNSAFE)
-{
-    FCALL_CONTRACT;
-
-    OBJECTREF obj = ObjectToOBJECTREF(objUNSAFE);
+    _ASSERTE((SIZE_T)flattenedIndex < refThis->GetNumComponents());
 
-    TypeHandle thTarget(target->type);
+    MethodTable* pElementTypeMT = arrayElementType.GetMethodTable();
+    PREFIX_ASSUME(NULL != pElementTypeMT);
 
-    MethodTable* pTargetMT = thTarget.AsMethodTable();
-    PREFIX_ASSUME(NULL != pTargetMT);
+    void* pData = refThis->GetDataPtr() + flattenedIndex * refThis->GetComponentSize();
 
     if (obj == NULL)
     {
         // Null is the universal zero...
-        if (pTargetMT->IsValueType())
-            InitValueClass(target->data,pTargetMT);
+        if (pElementTypeMT->IsValueType())
+            InitValueClass(pData,pElementTypeMT);
         else
-            ClearObjectReference((OBJECTREF*)target->data);
+            ClearObjectReference((OBJECTREF*)pData);
     }
     else
-    if (thTarget == TypeHandle(g_pObjectClass))
+    if (arrayElementType == TypeHandle(g_pObjectClass))
     {
         // Everything is compatible with Object
-        SetObjectReference((OBJECTREF*)target->data,(OBJECTREF)obj);
+        SetObjectReference((OBJECTREF*)pData,(OBJECTREF)obj);
     }
     else
-    if (!pTargetMT->IsValueType())
+    if (!pElementTypeMT->IsValueType())
     {
-        if (ObjIsInstanceOfCached(OBJECTREFToObject(obj), thTarget) != TypeHandle::CanCast)
+        if (ObjIsInstanceOfCached(OBJECTREFToObject(obj), arrayElementType) != TypeHandle::CanCast)
         {
-            // target->data is protected by the caller
-            HELPER_METHOD_FRAME_BEGIN_1(obj);
+            HELPER_METHOD_FRAME_BEGIN_2(refThis, obj);
 
-            if (!ObjIsInstanceOf(OBJECTREFToObject(obj), thTarget))
+            if (!ObjIsInstanceOf(OBJECTREFToObject(obj), arrayElementType))
                 COMPlusThrow(kInvalidCastException,W("InvalidCast_StoreArrayElement"));
 
             HELPER_METHOD_FRAME_END();
         }
 
-        SetObjectReference((OBJECTREF*)target->data,obj);
+        // Refresh pData in case GC moved objects around
+        pData = refThis->GetDataPtr() + flattenedIndex * refThis->GetComponentSize();
```

Comment:
Good catch. Out of curiosity, do you know if the compiler persists the precomputed value of `flattenedIndex * refThis->GetComponentSize()` across the call? I'd assume not since the compiler must assume that _ObjIsInstanceOf_ may change the component size.

I don't think this is _that_ perf-constrained that we need to kill the extra multiplication, but it piqued my curiosity.

Reply:
I have moved this up so that it is done for the super slow path only now.